keylime/allowlistdbproxy.py

```diff
@@ -0,0 +1,198 @@
+import threading
+import time
+
+from sqlalchemy import select
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.orm.exc import NoResultFound
+
+from keylime import keylime_logging
+from keylime.db.verifier_db import VerifierAllowlist
+
+logger = keylime_logging.init_logging("allowlistdbproxy")
+
+
+class AllowlistEntry:
+    """AllowlistEntry represents a named allowlist in the DB"""
+
+    @staticmethod
+    def from_db(session, allowlist_id):
+        """Create an AllowlistEntry object from an entry in the DB"""
+        try:
+            verifier_allowlist_db = session.query(VerifierAllowlist).filter_by(id=allowlist_id).one()
+        except NoResultFound:
+            return None
+        except SQLAlchemyError as e:
+            logger.error("SQLAlchemy Error: %s", e)
+            return None
+        return AllowlistEntry(session, allowlist_id, verifier_allowlist_db.checksum, verifier_allowlist_db)
+
+    def __init__(self, session, allowlist_id, checksum, verifier_allowlist_db, update_timeout=0):
+        """constructor"""
+        self.session = session
+
+        self.update_lock = threading.Lock()
+        self.last_update_check = time.monotonic()
+        self.update_timeout = update_timeout
+
+        self.allowlist_checksum = checksum
+        self.allowlist_id = allowlist_id
+        self.verifier_allowlist_db = verifier_allowlist_db
+        self.verifier_allowlist = AllowlistEntry.__copy_verifier_allowlist(self.verifier_allowlist_db)
+
+        # Use .where rather than .filter_by for older SQLAlchemy
+        self.query_checksum_stmt = select(VerifierAllowlist.checksum).where(VerifierAllowlist.id == self.allowlist_id)
+
+    @staticmethod
+    def __copy_verifier_allowlist(tocopy):
+        """Make a copy of the verifier_allowlist_db"""
+        # This is the object we will give to the user; it's basically a read-only object without
+        # access to the DB and should prevent that the user of the object sees changing fields
+        # while we refresh() the object from the DB
+        verifier_allowlist = VerifierAllowlist()
+        verifier_allowlist.id = tocopy.id
+        verifier_allowlist.name = tocopy.name
+        verifier_allowlist.ima_policy = tocopy.ima_policy
+        verifier_allowlist.tpm_policy = tocopy.tpm_policy
+
+        return verifier_allowlist
+
+    def __has_checksum_changed(self):
+        """Check whether the checksum of the allowlist in the DB has changed"""
+
+        verifier_allowlist = self.session.execute(self.query_checksum_stmt).one()
```

Comment:
This line here will put the biggest stress on the database but select only one field (checksum) for each allowlist.  If we have N clients this would query the table N times, once for each attestation, if no timeout is being used. If we use a timeout between queries, such as 0.1s for example, all those clients asking for the same allowlist within 0.1s would be bunched together after the first one queried for the checksum. This would reduce the stress on the DB. However, if during that 0.1s the allowlist is updated and an attestation is run on a system that just used a file that wasn't part of the old allowlist then we get an attestation failure. With no timeout we wouldn't risk any attestation failures... I am not sure how bad such a query for N clients possibly asking for N different entries in the table to get their checksums would be. Does anyone know?